File: poker/poker_player.cpp
#include "poker_player.h"

#include <future>
#include <vector>
#include <map>
#include <algorithm>

typedef std::map<PokerPlayer::PokerHand, std::future<float>> POKER_FUTURES;

PokerPlayer::PokerPlayer(const std::string& name, size_t player_num)
{
    playerName = name;
    playerNumber = player_num;
}


PokerPlayer::~PokerPlayer()
{
}


PokerPlayer::Action PokerPlayer::DeterminePlayerAction()
{
    //Determine if the hand has any poker hands in it and how strong that is.
    //Return an Action of FOLD, BET, RAISE, or CALL
    return (PokerPlayer::Action::FOLD);
}


PokerPlayer::PokerHand PokerPlayer::EvaluateHand()
{
   	FindPossibleHands();

    
    std::map<std::string, POKER_FUTURES> everything;
    std::map<std::string, PokerHand> highest_hands;

    for (CardCombo hand_set : possibleHands)
    {
      CARDS hand = GetPossibleHandasCARDS(hand_set);

      std::string handStr = DeckOfCards::CardsToString(hand);

    	float hand_results[11] = {0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0};

    	PokerHand highestHand = PokerHand::NOTHING;

    	everything[handStr][NOTHING] 			= (std::async(std::launch::async, [this, hand]{return IsNothing(hand);}));
    	everything[handStr][PAIR] 				= (std::async(std::launch::async, [this, hand]{return IsPair(hand);}));
    	everything[handStr][TWO_PAIR] 			= (std::async(std::launch::async, [this, hand]{return IsTwoPairs(hand);}));
    	everything[handStr][THREE_OF_A_KIND] 	= (std::async(std::launch::async, [this, hand]{return IsThreeOfaKind(hand);}));
    	everything[handStr][STRAIGHT] 			= (std::async(std::launch::async, [this, hand]{return IsStraight(hand);}));
    	everything[handStr][FLUSH] 				= (std::async(std::launch::async, [this, hand]{return IsFlush(hand);}));
    	everything[handStr][FULL_HOUSE] 		= (std::async(std::launch::async, [this, hand]{return IsFullHouse(hand);}));
    	everything[handStr][FOUR_OF_A_KIND] 	= (std::async(std::launch::async, [this, hand]{return IsFourOfaKind(hand);}));
    	everything[handStr][STRAIGHT_FLUSH] 	= (std::async(std::launch::async, [this, hand]{return IsStraightFlush(hand);}));
    	everything[handStr][ROYAL_FLUSH] 		= (std::async(std::launch::async, [this, hand]{return IsRoyalFlush(hand);}));
    	everything[handStr][FIVE_OF_A_KIND] 	= (std::async(std::launch::async, [this, hand]{return IsFiveOfaKind(hand);}));

    	// std::cout << "Is<HAND> functions have been started." << std::endl;

    	for (std::map<PokerHand, std::future<float> >::iterator it = everything[handStr].begin();
        	it != everything[handStr].end();
        	it++)
    	{
        	(*it).second.wait();

        	hand_results[(*it).first] = (*it).second.get();
        	// std::cout << "Hand score[" << HandNameToString((*it).first) << "] = " << hand_results[(*it).first] << std::endl;

        	if (1.0 == hand_results[(*it).first])
        	{
            		highestHand = (*it).first;
        	}
    	}    

      highest_hands[handStr] = highestHand;
    }

    std::string hand_cards;
    PokerHand ultimateHand = PokerHand::NOTHING;
    for (std::map<std::string, PokerHand>::iterator it = highest_hands.begin(); 
          it != highest_hands.end();
          it++)
    {
      if ( (*it).second > ultimateHand)
      {
        ultimateHand = (*it).second;
        hand_cards = (*it).first;
      } 
    }
    
    std::cout << "Ultimate hand: " << hand_cards;
    return (ultimateHand);
}


void PokerPlayer::ReceiveCard(Card card)
{
    playerHand.push_back(card);
    // std::cout << playerName << " receiving " << card.toString() 
    //           << ". HandSize = " << playerHand.size() << std::endl;
}


void PokerPlayer::ReceiveHand(CARDS hand)
{
    playerHand = hand;
}


CARDS PokerPlayer::ReturnCardsToDealer()
{
    CARDS oldHand = playerHand;
    playerHand.clear();

    return (oldHand);
}


CARDS PokerPlayer::GetPossibleHandasCARDS(CardCombo& card_set)
{
	CARDS hand;
	
	for (Card card : card_set)
	{
		hand.push_back(card);
	}

	return hand;
}


void PokerPlayer::FindPossibleHands()
{
    if ( 5 == playerHand.size() )
    {
      CardCombo combo;
      for (Card card : playerHand)
      {
        combo.insert(card);
      }

    	possibleHands.insert(combo);
      return;
    }
    CARDS allCards = playerHand;
    allCards.insert(allCards.end(), tableCards.cbegin(), tableCards.cend());

    //std::cout << "7 cards: " << std::endl;
    //DeckOfCards::PrintCards(allCards);

    do
    {
    	CardCombo combo;
    	for (int i=0; i<5; i++)
    	{
       		combo.insert(allCards[i]);
    	}

    	possibleHands.insert(combo);

    } while (std::next_permutation(allCards.begin(), allCards.end()));

//    std::cout << "Possible Hands: " << possibleHands.size() << std::endl;
}


std::string PokerPlayer::HandToString()
{
    std::string output = "[";
    // std::cout << "HandToString: handSize = " << playerHand.size() << std::endl;

    for (Card card : playerHand)
    {
        output += card.toString() + " ";
    }

    if (0 < tableCards.size())
    {
        output += "][";

        for (Card card : tableCards)
        {
            output += card.toString() + " ";
        }
    }

    output += "]";

    return (output);
}

std::string PokerPlayer::HandNameToString(PokerHand hand)
{
    std::string name;
    switch (hand)
    {
        case PokerHand::NOTHING:
            name = "Nothing";
            break;
        case PokerHand::PAIR:
            name = "Pair";
            break;
        case PokerHand::TWO_PAIR:
            name = "Two Pairs";
            break;
        case PokerHand::THREE_OF_A_KIND:
            name = "Three of a Kind";
            break;
        case PokerHand::STRAIGHT:
            name = "Straight";
            break;
        case PokerHand::FLUSH:
            name = "Flush";
            break;
        case PokerHand::FULL_HOUSE:
            name = "Full House";
            break;
        case PokerHand::FOUR_OF_A_KIND:
            name = "Four of a Kind";
            break;
        case PokerHand::STRAIGHT_FLUSH:
            name = "Straight Flush";
            break;
        case PokerHand::ROYAL_FLUSH:
            name = "Royal Flush";
            break;
        case PokerHand::FIVE_OF_A_KIND:
            name = "Five of a Kind";
            break;
        default:
            name = "Nothing (default)";
            break;

    }

    return (name);
}

float PokerPlayer::IsNothing(CARDS cards)
{
    float confidence = 1.0;
    return (confidence);
}

float PokerPlayer::IsPair(CARDS cards)
{
    float confidence = 0.0;

    std::map<std::string, size_t> value_counts;

    for (Card& card : cards)
    {
        value_counts[card.Value] += 1;
    }

    for (std::map<std::string, size_t>::iterator it = value_counts.begin();
         it != value_counts.end();
         it++)
    {
        if ((*it).second >= 2)
        {
            confidence = 1.0;
        }
    }
    return (confidence);
}

float PokerPlayer::IsTwoPairs(CARDS cards)
{
    float confidence = 0.0;

    std::map<std::string, size_t> value_counts;

    for (Card& card : cards)
    {
        value_counts[card.Value] += 1;
    }

    for (std::map<std::string, size_t>::iterator it = value_counts.begin();
         it != value_counts.end();
         it++)
    {
        if ((*it).second >= 2)
        {
            confidence += 0.5;
        }
    }
    return (confidence);
}

float PokerPlayer::IsThreeOfaKind(CARDS cards)
{
    float confidence = 0.0;
 
    std::map<std::string, size_t> value_counts;

    for (Card& card : cards)
    {
        value_counts[card.Value] += 1;
    }

    for (std::map<std::string, size_t>::iterator it = value_counts.begin();
         it != value_counts.end();
         it++)
    {
        if ((*it).second >= 3)
        {
            confidence = 1.0;
        }
    }
    return (confidence);
}

float PokerPlayer::IsStraight(CARDS cards)
{
    float confidence = 0.0;
    float highest_confidence = 0.0;
 
    std::map<std::string, size_t> value_counts;

    for (int i=0; i<DeckOfCards::NUM_VALUES; i++)
    {
        value_counts[VALUES[i]] = 0;
    }
    for (Card& card : cards)
    {
        value_counts[card.Value] += 1;
    }

    for (size_t i=0; i<DeckOfCards::NUM_VALUES && 1.0 != confidence; i++)
    {
        std::string value = VALUES[i];
        size_t value_count = value_counts[value];

        // std::cout << "Value[" << value << "] = " << value_count << std::endl;

        switch (value_count)
        {
            case 0:
                confidence = 0.0;
                break;
            case 1:
                confidence += 0.2;
                if (confidence > highest_confidence)
                {
                    highest_confidence = confidence;
                }
                // std::cout << "confidence = " << confidence << std::endl;
                break;
            default:
                confidence = 0.0;
                break;
        }
    }

    return (highest_confidence);
}

float PokerPlayer::IsFlush(CARDS cards)
{
    float confidence = 0.0;
 
    std::map<std::string, size_t> suit_counts;

    for (Card& card : cards)
    {
        suit_counts[card.Suit] += 1;
    }

    for (std::map<std::string, size_t>::iterator it = suit_counts.begin();
         it != suit_counts.end();
         it++)
    {
        if ((*it).second == 5)
        {
            confidence = 1.0;
        }
        else if ((*it).second >=3)
        {
            confidence = 0.2 * (*it).second;
        }
    }
    return (confidence);
}

float PokerPlayer::IsFullHouse(CARDS cards)
{
    float confidence = 0.0;
 
    std::map<std::string, size_t> value_counts;

    for (Card& card : cards)
    {
        value_counts[card.Value] += 1;
    }

    bool pairFound = false;

    for (std::map<std::string, size_t>::iterator it = value_counts.begin();
         it != value_counts.end();
         it++)
    {
        if ((*it).second == 3)
        {
            confidence += 0.6;
        }

        if ((*it).second == 2 && !pairFound)
        {
            pairFound = true;
            confidence += 0.4;
        }
    }
    return (confidence);
}

float PokerPlayer::IsFourOfaKind(CARDS cards)
{
    float confidence = 0.0;
 
    std::map<std::string, size_t> value_counts;

    for (Card& card : cards)
    {
        value_counts[card.Value] += 1;
    }

    for (std::map<std::string, size_t>::iterator it = value_counts.begin();
         it != value_counts.end();
         it++)
    {
        if ((*it).second >= 4)
        {
            confidence = 1.0;
        }
    }

    return (confidence);
}

float PokerPlayer::IsStraightFlush(CARDS cards)
{
    float confidence = 0.0;

    if (1.0 == IsStraight(cards))
    {
        confidence += 0.5;
    }

    if (1.0 == IsFlush(cards))
    {
        confidence += 0.5;
    }

    return (confidence);
}

float PokerPlayer::IsRoyalFlush(CARDS cards)
{
    float confidence = 0.0;

    if (1.0 == IsStraightFlush(cards))
    {
        confidence += 0.5;

        for (Card& card : cards)
        {
            if (VALUES[ACE] == card.Value)
            {
                confidence += 0.5;
                break;
            }
        }
    }

    return (confidence);
}

float PokerPlayer::IsFiveOfaKind(CARDS cards)
{
    float confidence = 0.0;
 
    std::map<std::string, size_t> value_counts;

    for (Card& card : cards)
    {
        value_counts[card.Value] += 1;
    }

    for (std::map<std::string, size_t>::iterator it = value_counts.begin();
         it != value_counts.end();
         it++)
    {
        if ((*it).second >= 5)
        {
            confidence = 1.0;
        }
    }

    return (confidence);
}

File: poker/poker_player.py
import asyncio
from enum import Enum
import copy
from deck_of_cards import VALUES, VALUE_IDS
from collections import defaultdict
import itertools

class Action(Enum):
    FOLD = 0
    BET = 1
    RAISE = 2
    CALL = 3

class PokerHand(Enum):
    NOTHING = 0
    PAIR = 1
    TWO_PAIR = 2
    THREE_OF_A_KIND = 3
    STRAIGHT = 4
    FLUSH = 5
    FULL_HOUSE = 6
    FOUR_OF_A_KIND = 7
    STRAIGHT_FLUSH = 8
    ROYAL_FLUSH = 9

POKER_HANDS = {
        PokerHand.NOTHING : 'Nothing',
        PokerHand.PAIR : 'Pair',
        PokerHand.TWO_PAIR : 'Two Pair',
        PokerHand.THREE_OF_A_KIND : 'Three of a Kind',
        PokerHand.STRAIGHT : 'Straight',
        PokerHand.FLUSH : 'Flush',
        PokerHand.FULL_HOUSE : 'Full House',
        PokerHand.FOUR_OF_A_KIND : 'Four of a Kind',
        PokerHand.STRAIGHT_FLUSH : 'Straight Flush',
        PokerHand.ROYAL_FLUSH : 'Royal Flush'
        }

class PokerPlayer:
    def __init__(self, new_name, new_number):
        self.name = new_name
        self.number = new_number
        self.hand = []

        # print(f'{self.name} {self.number} sitting down at the table.')

    def determine_player_action(self):
        return Action.FOLD

    def receive_hand(self, new_hand):
        self.hand = new_hand

    def receive_card(self, new_card):
        self.hand.append(new_card)
        # print(f'{self.name}{self.number} getting {new_card.to_string()}, have {len(self.hand)} cards.')

    def return_cards_to_dealer(self):
        old_hand = copy.deep_copy(self.hand) 
        self.hand.clear()
        return old_hand

    def get_name(self):
        return self.name

    def get_number(self):
        return self.number

    def get_hand(self):
        return self.hand

    def hand_to_string(self, hand):
        hand_as_string = ''
        for card in hand:
            hand_as_string += f'{card.to_string()} '

        return hand_as_string

    def hand_name_to_string(self, hand_name):
        return POKER_HANDS[hand_name]

    async def _is_nothing(self, hand):
        return 1.0

    async def _is_pair(self, hand):
        confidence = 0.0

        value_counts = defaultdict(int)

        for card in hand:
            value_counts[card.value] += 1

        for key,val in value_counts.items():
            if val >= 2:
                confidence = 1.0

        return confidence

    async def _is_two_pair(self, hand):
        confidence = 0.0

        value_counts = {}

        for card in hand:
            value_counts.setdefault(card.value, 0)
            value_counts[card.value] += 1

        for key,val in value_counts.items():
            if val >= 2:
                confidence += 0.5

            if confidence == 1.0:
                break

        return confidence

    async def _is_three_of_a_kind(self, hand):
        confidence = 0.0

        value_counts = defaultdict(int)

        for card in hand:
            value_counts[card.value] += 1

        for key,val in value_counts.items():
            if val >= 3:
                confidence = 1.0

        return confidence

    async def _is_straight(self, hand):
        confidence = 0.0
        highest_confidence = 0.0

        value_counts = defaultdict(int)

        for card in hand:
            value_counts[card.value] += 1

        for value in VALUES.values():
            if 1 == value_counts[value]:
                confidence += 0.2
                if confidence > highest_confidence:
                    highest_confidence = confidence
            else:
                confidence = 0.0

        return highest_confidence

    async def _is_flush(self, hand):
        confidence = 0.0

        suit_counts = defaultdict(int)

        for card in hand:
            suit_counts[card.suit] += 1

        for suit_count in suit_counts.values():
            if 5 == suit_count:
                confidence = 1.0
            #     break
            # elif 3 <= suit_count:
            #     confidence = 0.2 * suit_count

        return confidence

    async def _is_full_house(self, hand):
        confidence = 0.0

        value_counts = defaultdict(int)

        for card in hand:
            value_counts[card.value] += 1

        pair_found = False

        for count in value_counts.values():
            if 3 == count:
                confidence  += 0.6
            if 2 == count and not pair_found:
                pair_found = True
                confidence += 0.4

        return confidence

    async def _is_four_of_a_kind(self, hand):
        confidence = 0.0

        value_counts = defaultdict(int)

        for card in hand:
            value_counts[card.value] += 1

        for val in value_counts.values():
            if val >= 4:
                confidence = 1.0

        return confidence

    async def _is_straight_flush(self, hand):
        confidence = 0.0

        if 1.0 == await self._is_flush():
            confidence += 0.5

        if 1.0 == await self._is_straight():
            confidence += 0.5

        return confidence

    async def _is_royal_flush(self, hand):
        confidence = 0.0

        if 1.0 == await self._is_straight_flush():
            confidence += 0.8

        for card in hand:
            if VALUES[VALUE_IDS.ACE] == card.value:
                confidence += 0.2
                break

        return confidence

    async def evaluate_hand(self):
        if len(self.hand) < 5:
            return (self.hand, PokerHand.NOTHING)
        elif len(self.hand) > 5:
            highest_hand = PokerHand.NOTHING
            highest_cards = {}

            # make sets of 5 cards and test them
            # then pick the highest hand from the sets
            hands = itertools.combinations(self.hand, 5)

            for current_hand in hands:
                # print(f'current hand: {self.hand_to_string(current_hand)}')
                hand_results = await asyncio.gather(self._is_nothing(current_hand),
                                                    self._is_pair(current_hand),
                                                    self._is_two_pair(current_hand),
                                                    self._is_three_of_a_kind(current_hand),
                                                    self._is_straight(current_hand),
                                                    self._is_flush(current_hand),
                                                    self._is_full_house(current_hand),
                                                    self._is_four_of_a_kind(current_hand),
                                                    self._is_straight_flush(current_hand),
                                                    self._is_royal_flush(current_hand),
                                                    return_exceptions=True)

                # print(f'evaluate_hand({self.hand_to_string()} ={hand_results})')
                highest_hand = PokerHand.NOTHING
                highest_cards[current_hand] = PokerHand.NOTHING
                i = 0
                for hand_value in POKER_HANDS.keys():
                    if 1.0 == hand_results[i]:
                        highest_hand = hand_value

                    i += 1

                highest_cards[current_hand] = highest_hand
                # print(f"Player {self.number}, hand[{self.hand_to_string(current_hand)}] = {self.hand_name_to_string(highest_cards[current_hand])}")

            hight_hand_key = ''
            highest_hand_value = PokerHand.NOTHING
            for hand_key, hand_value in highest_cards.items():
                if hand_value.value >= highest_hand_value.value:
                    highest_hand_value = hand_value
                    hight_hand_key = hand_key

            # print(f'Possible hands: {hands}')
            return (hight_hand_key, highest_hand_value)
        else:
            hand_results = await asyncio.gather(self._is_nothing(self.hand),
                    self._is_pair(self.hand),
                    self._is_two_pair(self.hand),
                    self._is_three_of_a_kind(self.hand),
                    self._is_straight(self.hand),
                    self._is_flush(self.hand),
                    self._is_full_house(self.hand),
                    self._is_four_of_a_kind(self.hand),
                    self._is_straight_flush(self.hand),
                    self._is_royal_flush(self.hand),
                    return_exceptions=True)

            # print(f'evaluate_hand({self.hand_to_string()} ={hand_results})')
            highest_hand = PokerHand.NOTHING

            i = 0
            for hand in POKER_HANDS.keys():
                if 1.0 == hand_results[i]:
                    highest_hand = hand

                i += 1

            return (self.hand, highest_hand)

        return (self.hand, PokerHand.NOTHING)
